califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -30,55 +30,163 @@
-using roothacks::TCAHelper;
-using roothacks::TypedCollection;
+using namespace std;
+
+struct califa_candidate
+{
+
+    Int_t motherId;
+    vector<Int_t> crystalList;
+    Double_t energy;
+    Double_t ns;
+    Double_t nf;
+    Double_t theta;
+    Double_t phi;
+    ULong64_t time;
+};
+
+void R3BCalifaCrystalCal2Cluster::SetRectangularWindow(Float_t thetaL, Float_t phiL)
+{
+
+    fWindowAlg = "Rectangular";
+
+    fThetaLimit = thetaL;
+    fPhiLimit = phiL;
+};
+
+void R3BCalifaCrystalCal2Cluster::SetRoundWindow(Double_t window)
+{
+
+    fWindowAlg = "Round";
+    fRoundWindow = window;
+};
+
+bool compareByEnergy(R3BCalifaCrystalCalData* a, R3BCalifaCrystalCalData* b) { return a->GetEnergy() > b->GetEnergy(); }
+
+bool isInside(vector<Int_t>& vec, Int_t cryId) { return find(vec.begin(), vec.end(), cryId) != vec.end(); }
+
+bool R3BCalifaCrystalCal2Cluster::InsideClusterWindow(TVector3 mother, TVector3 crystal)
+{
+
+    if (fWindowAlg == "Round")
+        return mother.Angle(crystal) <= fRoundWindow;
+
+    if (fWindowAlg == "Rectangular")
+        return (abs(mother.Theta() - crystal.Theta()) < fThetaLimit) && (abs(mother.Phi() - crystal.Phi()) < fPhiLimit);
+
+    else
+        return 0;
+}
+
+void RemoveUsedCrystals(vector<Int_t>& used,
+                        vector<R3BCalifaCrystalCalData*>& all,
+                        vector<R3BCalifaCrystalCalData*>& proton,
+                        vector<R3BCalifaCrystalCalData*>& gamma,
+                        vector<R3BCalifaCrystalCalData*>& saturated)
+{
+
+    for (int p = 0; p < used.size(); p++)
+    {
+        for (int s = 0; s < all.size(); s++)
+
+            if (all.at(s)->GetCrystalId() == used.at(p))
+                all.erase(all.begin() + s);
+    }
+
+    for (int p = 0; p < used.size(); p++)
+    {
+        for (int s = 0; s < gamma.size(); s++)
+
+            if (gamma.at(s)->GetCrystalId() == used.at(p))
+                gamma.erase(gamma.begin() + s);
+    }
+
+    for (int p = 0; p < used.size(); p++)
+    {
+        for (int s = 0; s < proton.size(); s++)
+
+            if (proton.at(s)->GetCrystalId() == used.at(p))
+                proton.erase(proton.begin() + s);
+    }
+
+    for (int p = 0; p < used.size(); p++)
+    {
+        for (int s = 0; s < saturated.size(); s++)
+
+            if (saturated.at(s)->GetCrystalId() == used.at(p))
+                saturated.erase(saturated.begin() + s);
+    }
+}
+
+void addCrystal2Cluster(struct califa_candidate* cluster,
+                        R3BCalifaCrystalCalData* crystalCal,
+                        string range,
+                        vector<Int_t>* usedCrystals,
+                        Int_t fTotalCrystals)
+{
+
+    cluster->energy += crystalCal->GetEnergy();
+    cluster->ns += crystalCal->GetNs();
+    cluster->nf += crystalCal->GetNf();
+    cluster->crystalList.push_back(crystalCal->GetCrystalId());
+
+    usedCrystals->push_back(crystalCal->GetCrystalId());
+
+    if (range == "gamma")
```

Comment:
Why do you use string to classify the gamma or proton ranges? Why do not use the same notation as clustertype?